llvm/lib/Transforms/Vectorize/LoopVectorize.cpp

```diff
@@ -5313,13 +5313,18 @@ LoopVectorizationCostModel::getUniformMemOpCost(Instruction *I,
   // VF.getKnownMinValue() - 1 from a scalable vector. This does not represent
   // the actual generated code, which involves extracting the last element of
   // a scalable vector where the lane to extract is unknown at compile time.
-  return TTI.getAddressComputationCost(PtrTy) +
-         TTI.getMemoryOpCost(Instruction::Store, ValTy, Alignment, AS,
-                             CostKind) +
-         (IsLoopInvariantStoreValue
-              ? 0
-              : TTI.getVectorInstrCost(Instruction::ExtractElement, VectorTy,
-                                       CostKind, VF.getKnownMinValue() - 1));
+  InstructionCost Cost =
+      TTI.getAddressComputationCost(PtrTy) +
+      TTI.getMemoryOpCost(Instruction::Store, ValTy, Alignment, AS, CostKind);
+  if (!IsLoopInvariantStoreValue) {
+    if (VF.isFixed())
+      Cost += TTI.getVectorInstrCost(Instruction::ExtractElement, VectorTy,
+                                     CostKind, VF.getKnownMinValue() - 1);
+    else
+      Cost += TTI.getReverseVectorInstrCost(Instruction::ExtractElement,
+                                            VectorTy, CostKind, 0);
+  }
```

Comment:
The reason for the new TTI function is so you don't need to write code like this but can instead write code that is independent of the vector type.

It would be better if for fixed length types the base implementation of the new function simply calls the existing one with a modified index.